extensions/amp-inline-gallery/1.0/thumbnails.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {BaseCarousel} from '../../amp-base-carousel/1.0/base-carousel';
+import {CarouselContext} from '../../amp-base-carousel/1.0/carousel-context';
+import {px} from '../../../src/style';
+import {toWin} from '../../../src/types';
+import {
+  useContext,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './thumbnails.jss';
+
+/**
+ * @param {!InlineGalleryDef.ThumbnailProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function Thumbnails({
+  aspectRatio,
+  children,
+  className = '',
+  loop = false,
+  ...rest
+}) {
+  const classes = useStyles();
+  const pointerFine = window.matchMedia('(pointer: fine)');
```

Comment:
Ideally it should be a local window, yes. But most important: the window APIs should not be called in the main rendering path. We need a side-effect for this. Maybe:

```
const [pointerFine, setPointerFine] = useState(false);
useEffect(() => {
  const win = ref.current.ownerDocument.defaultView;
  setPointerFine(win.matchMedia(...));
}, []);
```

Expect this to become somewhat common, so we can always wrap this up into a custom hook and call it like this (name is not great):

```
const pointerFine = useEffectCompute(/*default*/ false, () => ref....matchMedia(...), []);
```

It kind of looks like `useMemo` here but resolved via a side-effect and thus can access refs and all available APIs through that.

Reply:
Done. I noticed that we can add event listeners to the media query, but decided not to include one here since I'm not familiar with how likely `pointer: fine` is to change. If it were something like `min-width` I'd attach, but left things simple for now.

Reply:
Possible. But I don't think it's a goal for us at this time. So totally ok to skip.